src/ios/MultipeerConnectivity/THEMultipeerClient.m

```diff
@@ -0,0 +1,284 @@
+
+    THEMultipeerClientSession *session = (THEMultipeerClientSession *)p;
+
+    NSString *peerIdentifier = info[PEER_IDENTIFIER_KEY];
+    if (session && ([[session remotePeerID] hash] == [peerID hash]) && 
+        [peerIdentifier isEqualToString:[session remotePeerIdentifier]])
+    {
+      NSLog(@"client: Found existing peer: %@", peerIdentifier);
+
+      clientSession = session;
+      previouslyVisible = clientSession.visible;
+    }
+    else
+    {
+      NSLog(@"client: Found new peer: %@", peerIdentifier);
+
+      // We've found a new peer, create a new record
+      clientSession = [[THEMultipeerClientSession alloc] 
+                                    initWithLocalPeerID:_localPeerId 
+                                       withRemotePeerID:peerID 
+                               withRemotePeerIdentifier:peerIdentifier];
+    }
+
+    [clientSession setVisible:YES];
+    return clientSession;
+  }];
+
+  if (clientSession && previouslyVisible == NO)
+  {
+    // A new peer or one that has become visible again. Only
+    // contact delegate when the state changes (we get duplicates a lot)
+    if ([_multipeerSessionDelegate respondsToSelector:@selector(didFindPeerIdentifier:peerName:)])
+    {
+      [_multipeerSessionDelegate 
+        didFindPeerIdentifier:[clientSession remotePeerIdentifier] 
+                     peerName:info[PEER_NAME_KEY]];
+    }
+  }
+}
+
+// Notifies the delegate that a peer was lost.
+- (void)browser:(MCNearbyServiceBrowser *)browser
+       lostPeer:(MCPeerID *)peerID
+{
+  __block BOOL previouslyVisible = NO;
+  __block THEMultipeerClientSession *clientSession = nil;
+
+  // Update the peer's record under lock
+  [_clientSessions updateForPeerID:peerID
+                       updateBlock:^THEMultipeerPeerSession *(THEMultipeerPeerSession *p) {
+
+    clientSession = (THEMultipeerClientSession *)p;
+    assert([[clientSession remotePeerID] hash] == [peerID hash]);
+
+    previouslyVisible = clientSession.visible;
+
+    if (clientSession)
+    {
+      NSLog(@"client: lost peer: %@", [clientSession remotePeerIdentifier]);
+      // disconnect will clear up any networking resources we currently hold
+      [clientSession disconnect];
+      [clientSession setVisible:NO];
+    }
+
+    return clientSession;
+  }];
+    
+  if (clientSession)
+  {
+    if (previouslyVisible == YES)
+    {
+      // Let interested parties know we lost a peer, only do this on a state change
+      if ([_multipeerSessionDelegate respondsToSelector:@selector(didLosePeerIdentifier:)])
+      {
+        [_multipeerSessionDelegate didLosePeerIdentifier:[clientSession remotePeerIdentifier]];
+      }
+    }
+  }
+  else
+  {
+    // Probably shouldn't happen
+    NSLog(@"WARNING: lostPeer we didn't know about");
+  }
+}
+
+- (void)browser:(MCNearbyServiceBrowser *)browser didNotStartBrowsingForPeers:(NSError *)error
+{
+  NSLog(@"WARNING: didNotStartBrowsingForPeers");
```

Comment:
PeerID only valid for the session it was created in. Evidenced by the fact you don't discover your current self but you will discover your previous selfs (when you're starting and stopping a lot). You will always fail to connect to your previous selfs.

Raised an exception instead, log does indicate programmer error.